src/emu/epoc/src/services/akn/icon/icon.cpp

```diff
@@ -65,4 +64,66 @@ namespace eka2l1 {
         create_session<akn_icon_server_session>(&context);
         context.set_request_status(KErrNone);
     }
+
+    void akn_icon_server::retrieve_icon(service::ipc_context *ctx) {
+        std::optional<epoc::akn_icon_params> spec = ctx->get_arg_packed<epoc::akn_icon_params>(0);
+        std::optional<epoc::akn_icon_srv_return_data> ret = ctx->get_arg_packed<epoc::akn_icon_srv_return_data>(1);
+
+        if (!spec || !ret) {
+            ctx->set_request_status(KErrArgument);
+            return;
+        }
+
+        std::optional<epoc::akn_icon_srv_return_data> cached = find_cached_icon(spec.value());
+        if (!cached) {
+            eka2l1::vec2 size = spec->size;
+            fbsbitmap *bmp = fbss->create_bitmap(size, init_data.icon_mode);
+            fbsbitmap *mask = fbss->create_bitmap(size, init_data.icon_mask_mode);
+
+            if (!bmp || !mask) {
+                ctx->set_request_status(KErrNoMemory);
+                return;
+            }
+
+            ret->bitmap_handle = bmp->id;
+            ret->content_dim.x = size.x;
+            ret->content_dim.y = size.y;
+            ret->mask_handle = mask->id;
+
+            add_cached_icon(ret.value(), spec.value());
+        } else {
+            ret.emplace(cached.value());
+        }
+
+        ctx->write_arg_pkg(0, spec.value());
+        ctx->write_arg_pkg(1, ret.value());
+        ctx->set_request_status(KErrNone);
+    }
+
+    std::optional<epoc::akn_icon_srv_return_data> akn_icon_server::find_cached_icon(epoc::akn_icon_params spec) {
```

Comment:
Passing it as a const reference. When you do that it just copy the original struct, which is unneccessary when you don't do any modification to the copy.

```suggestion
    std::optional<epoc::akn_icon_srv_return_data> akn_icon_server::find_cached_icon(const epoc::akn_icon_params &spec) {
```